src/meili/viterbi_search.cc

```diff
@@ -408,9 +408,19 @@ StateId ViterbiSearch::SearchWinner(StateId::Time time) {
   const StateId::Time max_allowed_time = unreached_states_by_time.size() - 1;
   const auto target = std::min(time, max_allowed_time);
 
-  // Continue last search if possible
-  StateId::Time searched_time = IterativeSearch(target, false);
+  /**
```

Comment:
`max_allowed_time` is used to make sure `time` is not exceeding it (see `const auto target = std::min(time, max_allowed_time);`), it doesn't involve in the actual iterative search loop. So the comments below don't make sense to me (unless you meant `target` time?).

`searched_time = IterativeSearch(target, ....)`'s goal is to find `winner_by_time[target]`. It can not guarantee that it can find the goal at the first time due to breakages, however, it guarantees that  `winner_by_time[searched_time]` will be found at least and `search_time` is increasing despite breakages. Therefore we call it in the while loop until `search_time >= target`, when winner at `search_time` is found, then the winner at `target` will be found too.

Reply:
yes, what get me confused is the part that, if time > max_allowed_time, we are not going to find any winner since it is out of scope. but we trimmed it to max_allowed_time and still run the iterative search. Are we caching anything? why don't we return {} directly?

Reply:
when time is greater than max_allowed_time, the logic trimmed it to max_allowed_time, and execute iterative search. If there is no breakage, the search will go to max_allowed_time and return. However, we are sill not expecting the logic to find anything, since, time is out of scope. What we really did is to fill winner_states_ by forcing the search continue to max_allowed_time. Unless I missed some critical part.

Reply:
Thanks a lot for answering my questions.

Reply:
I think there is a guarantee that: after SearchWinner(time) winners from time 0 to `time` should be calculated and cached. If we `return {}` on `max_allowed_time > time` in the beginning, then the guarantee will lose. I can't remember where are depending on this guarantee though. It looks StateIterator is okay without this guarantee because it always calls SearchWinner in Next.